### ISSUE TITLE
MAJOR WARNING TO INSTAGRAM-SDK-PHP LIBRARY USERS

### ISSUE THREAD

**Issue by SteveJobzniak** [liamcottle/Instagram-SDK-PHP#32]
# WARNING

Every time you call `$instagram->login()`, the library generates a new random "Android device hardware ID" and logs in to Instagram.

Instagram counts the number of device IDs registering from every IP.

When you exceed their threshold, they will blacklist your IP as an "open proxy". After that, you will tarnish any account you log in with via your IP (each such account will have to do email/phone validation), and you will not be able to create new accounts via that IP.

The solution?

From day 1, you must use Instagram-SDK-PHP's session save/restore function, which saves and restores the device ID instead of constantly calling `login`.

I sent some example code to the author a few months ago via email, for use in the example docs. Figured I'd now make this warning public since all examples still use the old "login with a new device ID every time" code.

During development you will frequently login several times per minute as you try things, so it's vital that every developer knows that they **must** use session restore code to avoid accidentally blacklisting your own IP.

Good luck, everyone!

**Comment by romalytvynenko** [comment]
@SteveJobzniak, thank you for sharing this vital info!

Are there any another known issues for this problem (going to instagram's black list)?

**Comment by SteveJobzniak** [comment]
@romalytvynenko Happy to help others avoid this disaster.

And no, there's no other way to become blacklisted as an "open proxy". They automatically blacklist you as a proxy when they see too many "new device IDs" from a single IP.

But here's some more general advice:

- Never create more than 5 accounts, to avoid getting banned by Instagram. I only use 1. But it's common knowledge that Instagram often bans all accounts associated with an IP if that IP has created more than 5 accounts. Five or less is universally agreed as "should be safe". The less the better. Their banning is manual, and having many accounts is what flags you for "review and punishment".
- Limit the rate at which you call their APIs. They have AI on the server which expects "human-like" interaction speed. I.e. scrolling slowly, clicking, waiting a bit, reading, clicking something else, etc. If you just spam them with hundreds of API queries in a row, they will block you temporarily (their APIs will return empty responses) to force you to chill out. The solution is to have a sane sleep() timer between every API call. They count your number of API calls in a certain time period. Think like a human user with a finger and a touchscreen. Safe sleep timers that won't hit the rate limit are anything in the range of 6-10 seconds between calls.

**Comment by pyrkamarcin** [comment]
@SteveJobzniak 
Can you see my project: [https://github.com/pyrkamarcin/instaxer]https://github.com/pyrkamarcin/instaxer) ?

One login session lasts 10 minutes min. I think I took everything you said.

**Comment by romalytvynenko** [comment]
@SteveJobzniak Thank you! This explains a lot.

What about generating device-id that depends on username? I don't say about login every time (session is must have), but in case when user will be logged out, his try to login will generate same device ID.

**Comment by SteveJobzniak** [comment]
@pyrkamarcin No. That code you posted will get blacklisted. I will quote your code here:

https://github.com/pyrkamarcin/instaxer/blob/master/run.php#L8
```php
$instaxer = new \Instaxer\Instaxer($user1, $pass1, 10, 10);
```

https://github.com/pyrkamarcin/instaxer/blob/master/src/Instaxer/Instaxer.php
```php
    /**
     * Instaxer constructor.
     * @param string $user
     * @param string $password
     * @param int $counter
     * @param int $long
     * @throws \Exception
     */
    public function __construct(string $user, string $password, int $counter = 10, int $long = 2)
    {
        $this->instagram = new Instagram();
        $this->instagram->setVerifyPeer(false);
        $this->instagram->login($user, $password);
        $this->counter = $counter;
        $this->long = $long;
        $this->marker = 0;
    }
```

Your code is logging in (calling `instagram->login()`) every time the constructor runs.

Every call to `login` generates a new device ID and logs in to Instagram, which they will count and finally blacklist you.

**Comment by SteveJobzniak** [comment]
@romalytvynenko 

>What about generating device-id that depends on username? I don't say about login every time (session is must have), but in case when user will be logged out, his try to login will generate same device ID.

That is a very, very good idea. Changing the "device info" generation code to be deterministic (based on username and always arriving at the exact same device ID) in this library would be a great change. Maybe even enough to totally avoid the need to save/restore the session.

However, there are many things that need to be done. Look at the `login()` function:
https://github.com/liamcottle/Instagram-SDK-PHP/blob/master/src/Instagram.php#L438

It calls `setupAsNewDevice` every time:
```php
    /**
     * Setup this instance with a fresh GUID, UUID and Phone ID.
     */
    public function setupAsNewDevice(){
        $guidId = Uuid::uuid4()->toString();
        $phoneId = Uuid::uuid4()->toString();
        $rankToken = Uuid::uuid4()->toString();
        $googleAdId = Uuid::uuid4()->toString();
        $this->setGuid($guidId);
        $this->setUuid($guidId);
        $this->setPhoneId($phoneId);
        $this->setRankToken($rankToken);
        $this->setGoogleAdId($googleAdId);
    }
```

And then it calls the following, to base the device ID on the username+password:
```php
        $this->setDeviceId($this->generateDeviceId(md5($username.$password)));
```

But that code in turn isn't as "stable" as it seems. It doesn't just base it on md5 of the username and password. It also bases the device ID on the modification time of the folder the script is in:
```php
    /**
     * @param $seed
     * @return string
     */
    public function generateDeviceId($seed){
        $volatile_seed = filemtime(__DIR__);
        return sprintf("android-%s", substr(md5($seed.$volatile_seed), 16));
    }
```

And there is also a problem with all of THESE random IDs, which also makes each login look like it comes from a different phone:
```php
        $guidId = Uuid::uuid4()->toString();
        $phoneId = Uuid::uuid4()->toString();
        $rankToken = Uuid::uuid4()->toString();
        $googleAdId = Uuid::uuid4()->toString();
```

So a solution that involves a rewrite of the library, would be something like this:

1. Write some code that generates a UUID from an input string, and always gives the exact same UUID for the exact same input string.
2. To make each UUID different (since we need about 5 different UUIDs to generate a "device fingerprint"), the code would have to look something like `generateUUID($username, 'guidid')`, and it would have to do something like `md5($username . $type)`, meaning it should join the string of the username + the type of Uuid we want, so the string is something like "stevejobzniakguidid", and "stevejobzniakdeviceid" etc, meaning every type of UUID we need will get a different result since it is fed a different "seed string" for every UUID we generate. 
3. Change all UUID generators (device ID, guidID, phoneID, rankToken, googleAdId, and anything else) to stop using the current random Uuid library, and instead use this function.

If this rewrite is made, and the same username always generates the exact same UUIDs, then yes it should be save to always call login() without saving the session! I have no time for this rewrite otherwise I would gladly do it (I've contributed other code over the months but now I am busy).

**Comment by pyrkamarcin** [comment]
@SteveJobzniak 
I understand what's going on in this issue. 10 minutes resulted from multiplying the amount of loop time and time sleep(). As you say, it is not important.

**Comment by SteveJobzniak** [comment]
@romalytvynenko @pyrkamarcin Anyway, here is some code (lots of comments added for you), which shows how to resume a session instead of generating new device IDs every time you want to login:

```php
// Attempt to restore an existing session.
// This prevents hammering the account, and prevents lockouts.
// Because the site EXPECTS persistent Android app sessions.
// It's also WAY faster. Without this, it generates a new fake
// device ID and session every time the script runs...
// ...which causes Instagram to flag your IP as an "open proxy"!
$restoredFromSession = FALSE;
$sessionFile = "sessions/{$account['username']}.dat";
if( is_file( $sessionFile ) ) {
    try {
        $savedSession = @file_get_contents( $sessionFile );
        if( $savedSession !== FALSE ) {
            // Restore session. This function doesn't validate it,
            // and just directly sets some local variables without
            // notifying Instagram at all, so after this we must
            // check if we're actually logged in afterwards!
            // That's actually a good thing, because Instagram
            // won't see us as constantly resuming/logging in again,
            // since the "restore" is done completely locally.
            // So to them it's like we never left!
            $instagram->initFromSavedSession( $savedSession );

            // Now check if we are logged in, by querying the API.
            $currentUser = $instagram->getCurrentUserAccount();
            if( $currentUser->getUser()->getUsername() == $account['username'] ) {
                // Successfully restored session!
                $restoredFromSession = TRUE;
            } else {
                // Saved session has expired. Needs new login.
            }
        }
    } catch( Exception $e ) {
        // Failed to restore the session...
    }
}

// Log in and save the login session if this is a new (re-)login.
if( ! $restoredFromSession ) {
    // Login (throws an error if incorrect credentials)
    $instagram->login( $account['username'], $account['password'] );

    // Serialize the session into a JSON string.
    $savedSession = $instagram->saveSession();

    // Store the session on disk.
    $result = @file_put_contents( $sessionFile, $savedSession );
    if( ! $result ) {
        // Unable to write session to file...
        @unlink( $sessionFile ); // delete the file
    }
}
```
